sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,304 @@
+
+        #[snafu(display("No IP Address for Primary Interface: {:?}", interface))]
+        NoIpAddress { interface: InterfaceId },
+    }
+}
+
+pub(crate) use error::Error;
+#[cfg(net_backend = "systemd-networkd")]
+type Result<T> = std::result::Result<T, error::Error>;
+
+#[cfg(test)]
+#[cfg(net_backend = "systemd-networkd")]
+mod tests {
+    use super::*;
+    use std::fs;
+    use std::net::Ipv4Addr;
+    use std::path::Path;
+    use std::path::PathBuf;
+
+    fn test_data() -> PathBuf {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test_data")
+    }
+
+    fn networkd_config() -> PathBuf {
+        test_data().join("networkd")
+    }
+
+    fn read_output_file<P>(path: P) -> String
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let path_str = fs::read_to_string(path).unwrap();
+
+        path_str
+    }
+
+    // full deserialize test
+    #[test]
+    fn parse_qemu_status() {
+        let ok = networkd_config().join("qemu_networkctl_output.json");
+        let network_status_str = read_output_file(ok);
+        let status_output: String = String::from_utf8(network_status_str.into()).unwrap();
+
+        // Parses correctly
+        let network_status_result = serde_json::from_str::<NetworkdStatus>(&status_output);
+        assert!(network_status_result.is_ok());
+
+        // Primary address is correct for file
+        let network_status = network_status_result.unwrap();
+        assert_eq!(
+            network_status.primary_address().unwrap(),
+            Ipv4Addr::new(10, 0, 2, 15)
+        );
+
+        if let Some(nameservers) = &network_status.dns {
+            let nameserver: Vec<IpAddr> = nameservers.iter().map(|n| n.address).collect();
+            assert_eq!(nameserver[0], Ipv4Addr::new(10, 0, 2, 3));
+        } else {
+            panic!("Nameservers not found in qemu_networkctl_output.json")
+        };
+
+        assert!(network_status.search_domains.is_none());
+    }
+
+    #[test]
+    fn parse_full_status() {
+        let ok = networkd_config().join("full_networkctl_output.json");
+        let network_status_str = read_output_file(ok);
+        let status_output: String = String::from_utf8(network_status_str.into()).unwrap();
+
+        // Parses correctly
+        let network_status_result = serde_json::from_str::<NetworkdStatus>(&status_output);
+        assert!(network_status_result.is_ok());
+
+        // Primary address is correct for file
+        let network_status = network_status_result.unwrap();
+        assert_eq!(
+            network_status.primary_address().unwrap(),
+            Ipv4Addr::new(172, 31, 28, 92)
+        );
+
+        if let Some(nameservers) = &network_status.dns {
+            let nameserver: Vec<IpAddr> = nameservers.into_iter().map(|n| n.address).collect();
+            assert_eq!(nameserver[0], Ipv4Addr::new(172, 31, 0, 2));
+        } else {
+            panic!("Nameservers not found in qemu_networkctl_output.json")
```

Comment:
This message references the wrong file: `qemu...`

Reply:
These names need work, I think I should shift them around a bit anyway, so I'll refactor them into better names that bring some meaning to their purpose rather than their origin.